price-feeder/oracle/oracle.go

```diff
@@ -33,15 +66,281 @@ func (o *Oracle) Stop() {
+		}
+
+		g.Go(func() error {
+			var ticker []string
+			for _, cp := range currencyPairs {
+				ticker = append(ticker, cp.String())
+			}
+
+			prices, err := priceProvider.GetTickerPrices(ticker...)
+			if err != nil {
+				return err
+			}
+
+			// flatten and collect prices based on the base currency per provider
+			//
+			// e.g.: {ProviderKraken: {"ATOM": 34.03, "UMEE": 6.3}}
+			mtx.Lock()
+			for _, cp := range currencyPairs {
+				if _, ok := providerPrices[providerName]; !ok {
+					providerPrices[providerName] = make(map[string]sdk.Dec)
+				}
+				if p, ok := prices[cp.String()]; ok {
+					providerPrices[providerName][cp.Base] = p
+				}
+			}
+			mtx.Unlock()
+
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
+	// consolidate the different provider maps into one for each exchange rate
+	var (
+		priceAverages = make(map[string]sdk.Dec)
+		priceCounts   = make(map[string]int)
+	)
+
+	// TODO: Consider using Volume Weighted Average Price (VWAP).
+	//
+	// Ref: https://github.com/umee-network/umee/issues/251
+	for _, prices := range providerPrices {
+		for base, price := range prices {
+			if _, ok := priceAverages[base]; !ok {
+				priceAverages[base] = sdk.NewDec(0)
+			}
+			priceAverages[base] = priceAverages[base].Add(price)
+			priceCounts[base]++
+		}
+	}
+
+	for k, average := range priceAverages {
+		average = average.QuoInt64(int64(priceCounts[k]))
+	}
+
+	o.prices = priceAverages
+
+	return nil
+}
+
+func (o *Oracle) GetParams() (*umeetypes.QueryParamsResponse, error) {
+	// Create a connection to the gRPC server.
+	grpcConn, err := grpc.Dial(
+		o.oracleClient.GRPCEndpoint, // your gRPC server address.
+		grpc.WithInsecure(),         // The Cosmos SDK doesn't support any transport security mechanism.
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	defer grpcConn.Close()
+	queryClient := umeetypes.NewQueryClient(grpcConn)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	queryResponse, err := queryClient.Params(ctx, &umeetypes.QueryParamsRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	return queryResponse, nil
+}
+
+func (o *Oracle) generateSalt(length int) (string, error) {
+	if length == 0 {
+		return "", fmt.Errorf("can't generate empty salt")
+	}
+	n := length / 2
+	bytes := make([]byte, n)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(bytes), nil
+}
+
+var previousVotePeriod float64
+
+type PreviousPrevote struct {
+	ExchangeRates     string
+	Salt              string
+	SubmitBlockHeight int64
+}
+
+func NewPreviousPrevote() *PreviousPrevote {
+	return &PreviousPrevote{
+		Salt:              "",
+		ExchangeRates:     "",
+		SubmitBlockHeight: 0,
+	}
+}
+
+var previousPrevote *PreviousPrevote = nil
+
+func (o *Oracle) tick() error {
+	ctx, err := o.oracleClient.CreateContext()
+	if err != nil {
+		return err
+	}
+
+	err = o.SetPrices()
+	if err != nil {
```

Comment:
just a bit more tidy :P 

```suggestion
	if err := o.SetPrices(); err != nil {
```